kodein-di/src/commonTest/kotlin/org/kodein/di/Tests_28_ParameterizedNew.kt

```diff
@@ -0,0 +1,107 @@
+package org.kodein.di
+
+import org.kodein.di.test.FixMethodOrder
+import org.kodein.di.test.MethodSorters
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+import kotlin.test.assertNotSame
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+internal class Tests_28_ParameterizedNew {
+
+    open class A(val i: Int = 0)
+    class B(val p: A)
+
+    @Test
+    fun test_00_new_1param() {
+        val di = DI {
+            bind { singleton { A() } }
+            bind { factory { p: A -> new(p, ::B) } }
+        }
+
+        val arg = A()
+        val b: B by di.instance(arg = arg)
+
+        assertEquals(b.p, arg, "expected parameter, but injected instance")
+    }
+
+    class C(val p1: A, val p2: A)
+
+    @Test
+    fun test_01_new_2param() {
+        val di = DI {
+            bind { singleton { A() } }
+            bind { factory { p: A -> new(p, ::C) } }
+        }
+
+        val injected by di.instance<A>()
+
+        val param = A()
+        val c: C by di.instance(arg = param)
+
+        assertEquals(c.p1, param, "expected parameter, but injected instance")
+        assertEquals(c.p2, injected, "expected instance, but injected parameter")
+    }
+
+    @Test
+    fun test_03_new_2param_no_injected() {
+        val di = DI {
+            //            bind { singleton { A() } } // on purpose
+            bind { factory { p: A -> new(p, ::C) } }
+        }
+
+        assertFailsWith<DI.NotFoundException> {
+            val param = A()
+            val c: C by di.instance(arg = param) // first param succeeds, second fails
+            println("p1=${c.p1} p2=${c.p2}")
+        }
+    }
+
+
+    class S() : A()
+
+    @Test
+    fun test_02_new_subclass() {
+        val di = DI {
+            bind { singleton { A() } }
+            bind { factory { p: A -> new(p, ::B) } }
+        }
+
+        val subclass = S()
+        val b: B by di.instance(arg = subclass)
+
+        assertEquals(b.p, subclass, "injected parameter ${b.p} is not the expected subclass")
+    }
+
+    class I
+    class D(val p: I)
+
+    @Test
+    fun test_03_new_nousage() {
+        val i = I()
+        val di = DI {
+            bind { singleton { i } }
+            bind { factory { p: A -> new(p, ::D) } }
+        }
+
+        val param = A() // injected class does not use the param
```

Comment:
Here, this test passes even though we never used the parameter at all. It was not used in the injected dependency (`D`) constructor. 

So the parameter `A` that the client provided is wasted.

This can lead to unintended behavior if the developer made a mistake of not including parameters, but accidentally used another class, which happened to also be present in the DI graph. 

However, it's not possible to fix this in compile time, and the only solution I see is to throw something like `DI.UnusedParameterException` after creating an instance.

Suggestion: Choose one option: 

1. Leave as is : more convenient and lenient, but can cause runtime bugs
2. Change to add verification code and throw: Strict, but will crash in runtime at instance's first usage.

Reply:
I'd rather see something more strict. I'd prefer have a runtime exception about an unused parameter, meaning my code is not properly designed, than having wired behavior from product perspective.

So let's go for 2. if you're ok with that.